_posts/2023-10-17-rust-in-yocto.md

```diff
@@ -0,0 +1,149 @@
+
+1. **The Original `meta-rust/meta-rust` Layer**: This method serves as a beacon for compatibility with older Yocto releases and furnishes a rust compiler for all Yocto versions. Nonetheless, it bears its own set of challenges:
+    - It mandates the inclusion of an additional layer to your project.
+    - The process of rebuilding the rust compiler from source can considerably slow down the build.
+    - Utilizing `cargo bitbake` is essential for recipe preparation, and this demands upkeep, especially when there are alterations in the `Cargo.lock` dependencies.
+
+2. **The `openembedded` Rust Class**: Functioning as the official Rust endorsement for Yocto, it's seamlessly integrated into open-oe. However, certain limitations might overshadow its benefits:
+    - Rust versions are firmly anchored to specific Yocto releases, which might render them outdated for some projects.
+    - Reconstructing the rust compiler from its source can be a time-consuming affair.
+    - Maintaining the recipe becomes inevitable, more so with the necessity to enumerate the complete list of dependencies.
+
+3. **The `meta-rust-bin/meta-rust` Recipe**: This alternative champions the use of pre-constructed binaries of rust, cargo, and libstd. And the recipes are succinct, eliminating the need to itemize all dependencies. Nevertheless, it has its nuances:
+    - It introduces an additional layer to your project.
+    - Its recipe blueprint diverges from the `meta-rust` recipes.
+
+At Memfault, our inclination leans towards the `meta-rust-bin`. Its compatibility with a broader range of Yocto versions combined with its efficiency in reducing build time strikes a chord with us. That said, drafting a recipe for `meta-rust` isn't a complex undertaking, especially with the assistance of `cargo bitbake`. We keep one at our disposal, catering to clients who favor `meta-rust`.
+
+## Another Important Consideration: Binaries Size
+
+When delving into Rust for embedded systems, certain facets remain untouched in our discussion but merit attention in future deliberations.
+
+- **The Rust Standard Library**: A conspicuous drawback in the methods discussed is the static linking of the Rust standard library into each binary file. This addition can augment the file by several hundred kilobytes. An ingenious workaround might be the construction of `libstd-rs` as a dynamic library, allowing all the rust binaries within the system to share it. As Rust binaries become an integral part of your system image, this factor will gain prominence. None of the solutions presented today support this but they all mention it as a future improvement. Presently, our tactic involves dynamically linking (filesystem links) our three binaries (`memfaultd`, `memfaultctl`, and `memfault-core-handler`) to a singular binary. This unique binary alters its function based on the name used in the invocation, a strategy we fondly refer to as the `busybox` approach.
```

Comment:
```suggestion
- **The Rust Standard Library**: A conspicuous drawback in the methods discussed is the static linking of the Rust standard library into each binary file. This addition can augment the file by several hundred kilobytes. An ingenious workaround might be the construction of `libstd-rs` as a dynamic library, allowing all the rust binaries within the system to share it. As Rust binaries become an integral part of your system image, this factor will gain prominence. None of the solutions presented today support this but they all mention it as a future improvement. Presently, our tactic involves symlinking (filesystem links) our three programs (`memfaultd`, `memfaultctl`, and `memfault-core-handler`) to a singular binary. This unique binary alters its function based on the name used in the invocation, a strategy we fondly refer to as the `busybox` approach.
```

I wouldn't use "dynamically linking" there.

Reply:
Same idea, as the webinar callout above but for recommended readings of previous posts. 

>📚**Recommended Reading**
>
>Fan of Rust? Check out our other Rust articles: [Asynchronous Rust on Cortex-M Microcontrollers](https://interrupt.memfault.com/blog/embedded-async-rust), [Rust for Low Power Digital Signal Processing](https://interrupt.memfault.com/blog/rust-for-digital-signal-processing), [From Zero to main(): Bare metal Rust](https://interrupt.memfault.com/blog/zero-to-main-rust-1).